README.md
# Zea PointClouds

[![NPM Package][npm]][npm-url]
[![Build Size][build-size]][build-size-url]
[![NPM Downloads][npm-downloads]][npmtrends-url]

# Introduction
Zea Point Clouds is a plug-in for [Zea Engine](https://docs.zea.live/zea-engine) that enables large point clouds rendering. It is based on [Potree](https://github.com/potree/potree/) with modifications to integrate into the [Zea Engine](https://docs.zea.live/zea-engine) Framework.
</br>
</br>
</br>

# Documentation
Full documentation with concepts, tutorials, live examples, API documentation and more; can be found at the zea ux docs site:
[https://docs.zea.live/zea-pointclouds](https://docs.zea.live/zea-pointclouds)

These docs allow developers to get started with the Zea UX by downloading free and open-source demo content and using Zea's publicly distributed client-side libraries.
</br>
</br>
</br>

# Licensing
The Zea Potree in is under a [`MIT`](https://en.wikipedia.org/wiki/MIT_License) license.
</br>
</br>
</br>

# Add it to your project
The process to add Zea Potree to your projects is easy. 

## *Using CDNs*
For static websites or quick implementation you can always use CDNs like JsDelivr or Unpkg:

### *JsDelivr*
```html
<script crossorigin src="https://cdn.jsdelivr.net/npm/@zeainc/zea-potree/dist/index.umd.min.js"></script>
```
### *Unpkg*
```html
<script crossorigin src="https://unpkg.com/@zeainc/zea-potree/dist/index.umd.js"></script>
```
### *Use it*
```html
<script>
  const { PointCloudAsset } = globalThis.zeapointclouds
</script>
```

</br>
</br>
</br>

# Dependencies
This plug-in depends on [ZeaEngine](https://docs.zea.live/zea-engine) and [ZeaUX](https://docs.zea.live/zea-ux). So, if you're using CDNs, make sure to import these two libraries before pointclouds.
</br>


> For questions on licensing, please fill out the contact form on our website: [_zea.live_](https://www.zea.live/contact-us)

[npm]: https://badge.fury.io/js/%40zeainc%2Fzea-pointclouds.svg
[npm-url]: https://www.npmjs.com/package/@zeainc/zea-pointclouds
[build-size]: https://badgen.net/bundlephobia/minzip/@zeainc/zea-pointclouds
[build-size-url]: https://bundlephobia.com/result?p=@zeainc/zea-pointclouds
[npm-downloads]: https://img.shields.io/npm/dw/@zeainc/zea-pointclouds
[npmtrends-url]: https://www.npmtrends.com/@zeainc/zea-pointclouds


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [4.0.1](https://github.com/ZeaInc/zea-pointclouds/compare/v4.0.0...v4.0.1) (2022-02-07)


### Bug Fixes

* Move Engine to dev dependencies ([f1b3243](https://github.com/ZeaInc/zea-pointclouds/commit/f1b324335ad81c102b36042ae2e472f5cec55007))

## [4.0.0](https://github.com/ZeaInc/zea-potree/compare/v3.2.0...v4.0.0) (2021-11-12)


### ⚠ BREAKING CHANGES

* Migrated to zea-engine 4.0.0

### Bug Fixes

* Fixed issue preventing zea-potree from working with bundlers. ([2152742](https://github.com/ZeaInc/zea-potree/commit/2152742a8c10d279c6a67aea8f21c27055c13e1c))


* Migrated to zea-engine 4.0.0 ([4fcf948](https://github.com/ZeaInc/zea-potree/commit/4fcf948fdc184907813523f613ee6d84c3202cae))

### [3.2.2](https://github.com/ZeaInc/zea-potree/compare/v3.2.0...v3.2.2) (2021-11-12)


### Bug Fixes

* Fixed issue preventing zea-potree from working with bundlers. ([2152742](https://github.com/ZeaInc/zea-potree/commit/2152742a8c10d279c6a67aea8f21c27055c13e1c))

### [3.0.1] (2020-12-04)

## [3.0.0] (2020-12-04)


### ⚠ BREAKING CHANGES

* **npm:** Raw imports are no longer supported.

### Features

* Added documentation for point clouds ([390f0c9](https://github.com/ZeaInc/zea-pointclouds/commit/390f0c9e87d764ab15ffce45df4ffe64ee0d5b53))
* Added zea standardised documentation ([ccbad2c](https://github.com/ZeaInc/zea-pointclouds/commit/ccbad2c75b152492830a69805f6b1575487244b5))
* Eye-Dome lighting implemented. ([8a376c8](https://github.com/ZeaInc/zea-pointclouds/commit/8a376c8de4d46ade3dbd7550a396db11bee544f8))


### Bug Fixes

* Added download demo code and fixed favicon ([8fe8700](https://github.com/ZeaInc/zea-pointclouds/commit/8fe870080b1571ec47959f69262b5a153074cb89))
* Fixed styles for download link ([4b4b0df](https://github.com/ZeaInc/zea-pointclouds/commit/4b4b0df6e2454aa856df2d946f59f627d4a0cff9))
* Rename bug on docs ([cc8b4e0](https://github.com/ZeaInc/zea-pointclouds/commit/cc8b4e0966393eb798c6ae90187a94991b3f0742))
* Specified the folder to use when generate documentation ([2616d27](https://github.com/ZeaInc/zea-pointclouds/commit/2616d271d7970789b0295f972ff85cfe84cb45cc))
* updated docs ([8f32322](https://github.com/ZeaInc/zea-pointclouds/commit/8f32322570374e08fe286277ec68ab836ba60635))
* updated docs ([a645960](https://github.com/ZeaInc/zea-pointclouds/commit/a645960f188ad7ad1702fd2f49b0b7a95fe27e3c))


### build

* **npm:** Add UMD support ([#7](https://github.com/ZeaInc/zea-pointclouds/issues/7)) ([ec04a7b](https://github.com/ZeaInc/zea-pointclouds/commit/ec04a7bf439202b951f9039a59fff2bf28e0f61c)), closes [#8](https://github.com/ZeaInc/zea-pointclouds/issues/8)


rollup.config.js
import { terser } from 'rollup-plugin-terser'
import json from '@rollup/plugin-json'
import webWorkerLoader from 'rollup-plugin-web-worker-loader'

import pkg from './package.json'

const external = ['@zeainc/zea-engine']

const plugins = [json(), webWorkerLoader({ targetPlatform: 'browser' })]

const isProduction = !process.env.ROLLUP_WATCH

if (isProduction) {
  plugins.push(terser())
}

const sourcemap = true

export default [
  // Browser-friendly UMD build.
  {
    input: 'src/zea/index.js',
    external,
    output: {
      name: 'zeaPotree',
      file: pkg.browser,
      format: 'umd',
      sourcemap,
      globals: {
        '@zeainc/zea-engine': 'zeaEngine',
      },
    },
    plugins,
  },

  // CommonJS (for Node) and ES module (for bundlers) build.
  // (We could have three entries in the configuration array
  // instead of two, but it's quicker to generate multiple
  // builds from a single configuration where possible, using
  // an array for the `output` option, where we can specify
  // `file` and `format` for each target)
  {
    input: 'src/zea/index.js',
    external,
    output: [
      { file: pkg.main, format: 'cjs', sourcemap },
      { file: pkg.module, format: 'es', sourcemap },
    ],
    plugins,
  },
]


src/Zea/PointCloudAsset.js
import { Box3, NumberParameter, AssetItem, Registry } from '@zeainc/zea-engine'
import { POCLoader } from '../loader/POCLoader'

// Note: replaces PointCloudOctree.
//
/**
 * PointCloudAsset is an application of Tree Items with rendering and material capabilities.
 * In this case, it is an specification that supports Point clouds Octree geometries.
 *
 * **Parameters**
 * *  **Num Points(`NumberParameter`):** Sets the number of points in the octree geometry
 * * **Point Size(`NumberParameter`):** Sets the size of the individual points inside the geometry
 * * **Point Size Attenuation(`NumberParameter`):**
 *
 * **Events**
 * * **loaded:** Triggered after the asset is loaded
 *
 * @extends {AssetItem}
 */
class PointCloudAsset extends AssetItem {
  /**
   * Creates an instance of PointCloudAsset.
   */
  constructor() {
    super()

    this.numPointsParam = new NumberParameter('Num Points', 0)
    this.pointSizeParam = new NumberParameter('Point Size', 1)
    this.pointSizeAttenuationParam = new NumberParameter('Point Size Attenuation', 1.0)

    this.pointBudget = 5 * 1000 * 100
    this.minimumNodeVSize = 0.2 // Size, not in pixels, but a fraction of screen V height.
    this.level = 0
    this.visibleNodes = []

    this.loaded = false

    // this.fileParam = this.addParameter(new FilePathParameter('File'))
    // this.fileParam.on('valueChanged', () => {
    //   this.loaded.untoggle()
    //   this.loadPointCloud(path, name)
    // })
    // this.addParameter(new NumberParameter('Version', 0))
    this.addParameter(this.numPointsParam)
    this.addParameter(this.pointSizeParam)
    this.addParameter(this.pointSizeAttenuationParam)
  }

  /**
   * The getGlobalMat4 method
   *
   * @return {Mat4} - The global Mat4
   */
  getGlobalMat4() {
    return this.globalXfoParam.value.toMat4()
  }

  /**
   * The _cleanBoundingBox method
   *
   * @param {Box3} bbox -
   * @return {Box3} - The cleaned bounding box
   * @private
   */
  _cleanBoundingBox(bbox) {
    bbox = super._cleanBoundingBox(bbox)
    const mat4 = this.getGlobalMat4()
    const geomBox = new Box3()
    const { min, max } = this.pcoGeometry.tightBoundingBox
    geomBox.min.set(min.x, min.y, min.z)
    geomBox.max.set(max.x, max.y, max.z)
    bbox.addBox3(geomBox, mat4)
    return bbox
  }

  /**
   * Sets asset's Point Cloud Geometry data.
   *
   * @param {PointCloudOctreeGeometry} pcoGeometry - The pcoGeometry value
   */
  setGeometry(pcoGeometry) {
    this.pcoGeometry = pcoGeometry

    const xfo = this.globalXfoParam.value
    xfo.tr = this.pcoGeometry.offset
    this.globalXfoParam.value = xfo

    // this.getParameter('Version').setValue(parseFloat(pcoGeometry.version));
    if (pcoGeometry.numPoints) this.numPointsParam.value = pcoGeometry.numPoints

    // this._setBoundingBoxDirty()

    this.loaded = false
    this.emit('loaded')

    // if (this.viewport)
    //   this.updateVisibility();
  }

  /**
   * Returns asset's point cloud Octree geometry
   *
   * @return {PointCloudOctreeGeometry} - The pcoGeometry
   */
  getGeometry() {
    return this.pcoGeometry
  }

  /**
   * Loads a Point Cloud Octree Geometry from the specified url.
   *
   * @param {string} path - The path value
   * @param {string} name - The name value
   * @return {Promise} - The result
   */
  load(path) {
    return new Promise((resolve, reject) => {
      POCLoader.load(path, (geometry) => {
        if (!geometry) {
          reject(`failed to load point cloud from URL: ${path}`)
        } else {
          this.setGeometry(geometry)
          resolve(geometry)
        }
      })
    })
  }

  loadPointCloud(path, name) {
    return this.load(path)
  }
}

Registry.register('PointCloudAsset', PointCloudAsset)

export default PointCloudAsset
export { PointCloudAsset }


src/Zea/GLPointCloudPass.js
import { Frustum, NumberParameter, GLTexture2D, GLPass, GLRenderTarget, GLMesh, Plane, Color } from '@zeainc/zea-engine'

import { PointCloudAsset } from './PointCloudAsset.js'
import { GLPointCloudAsset } from './GLPointCloudAsset.js'
import { BinaryHeap } from '../../libs/other/BinaryHeap.js'
import { PointCloudShader, PointCloudGeomDataShader, PointCloudHilighlightShader } from './PointCloudShader.js'
import { LRU } from '../LRU.js'
import { EyeDomeLightingShader } from './EyeDomeLightingShader.js'
import { SmoothingShader } from './SmoothingShader.js'

/**
 * GLPointCloudPass abstracts the rendering of cloud point geometries to the screen.
 *
 * **Parameters**
 * **MinimumNodeVSize(`NumberParameter`)**
 * **VisiblePointsTarget(`NumberParameter`)**
 *
 * **Events**
 * * *updated:* Triggers every time one of the Point Cloud Asset changes.
 *
 * @extends {GLPass}
 */
class GLPointCloudPass extends GLPass {
  /**
   * Creates an instance of GLPointCloudPass.
   */
  constructor() {
    super()

    this.visiblePointsTarget = 2 * 1000 * 1000
    this.lru = new LRU()
    this.minimumNodeVSize = 0.2
    this.eyeDomeLighting = false
    this.glpointcloudAssets = []
    this.hilghlightedAssets = []

    this.visibleNodesNeedUpdating = false

    // Size, not in pixels, but a fraction of scnreen V height.
    const minimumNodeVSizeParam = this.addParameter(new NumberParameter('MinimumNodeVSize', 0.0))
    minimumNodeVSizeParam.on('valueChanged', () => {
      this.minimumNodeVSize = minimumNodeVSizeParam.getValue()
    })

    const visiblePointsTargetParam = this.addParameter(new NumberParameter('VisiblePointsTarget', 0))
    visiblePointsTargetParam.on('valueChanged', () => {
      this.pointBudget = visiblePointsTargetParam.getValue()
      this.lru.pointLoadLimit = this.pointBudget * 2
    })

    // const pointSizeParam = this.addParameter(new NumberParameter('Points Size', 0))
    // pointSizeParam.on('valueChanged', () => {
    //   this.pointSize = pointSizeParam.getValue()
    // });

    minimumNodeVSizeParam.setValue(0.2)
    visiblePointsTargetParam.setValue(2 * 1000 * 1000)
    // pointSizeParam.setValue(1.25)
  }

  /**
   * Initializes the rendering of the point cloud geometries hosted.
   *
   * @param {GLRenderer} renderer - The renderer param.
   * @param {number} passIndex - The passIndex param.
   */
  init(renderer, passIndex) {
    super.init(renderer, passIndex)
    const gl = renderer.gl
    this.glshader = new PointCloudShader(gl)
    this.glgeomdataShader = new PointCloudGeomDataShader(gl)
    this.glhighlightShader = new PointCloudHilighlightShader(gl)

    const size = 2048
    const data = new Uint8Array(size * 4)
    for (let i = 0; i < size * 4; i++) data[i] = 255

    this.visibleNodesTexture = new GLTexture2D(gl, {
      format: 'RGBA',
      type: 'UNSIGNED_BYTE',
      width: size,
      height: 1,
      filter: 'NEAREST',
      wrap: 'CLAMP_TO_EDGE',
      mipMapped: false,
      data,
    })

    this.setViewport(renderer.getViewport())
  }

  initEyeDomeLighting() {
    const gl = this.__gl

    this.primaryRenderTarget = new GLRenderTarget(gl, {
      type: 'FLOAT',
      format: 'RGBA',
      filter: 'NEAREST',
      width: this.viewportSize[0],
      height: this.viewportSize[1],
      depthType: gl.FLOAT,
      depthFormat: gl.DEPTH_COMPONENT,
      depthInternalFormat: gl.DEPTH_COMPONENT32F,
    })
    this.primaryRenderTarget.clearColor.a = 99999.0

    this.smoothedRenderTarget = new GLRenderTarget(gl, {
      type: 'FLOAT',
      format: 'RGBA',
      filter: 'NEAREST',
      width: this.viewportSize[0],
      height: this.viewportSize[1],
      depthType: gl.FLOAT,
      depthFormat: gl.DEPTH_COMPONENT,
      depthInternalFormat: gl.DEPTH_COMPONENT32F,
    })
    this.smoothedRenderTarget.clearColor.a = 99999.0

    this.quad = new GLMesh(gl, new Plane(1, 1))
    this.eyeDomeLightingShader = new EyeDomeLightingShader(gl)
    this.smoothingShader = new SmoothingShader(gl)
  }

  /**
   * The itemAddedToScene method is called on each pass when a new item
   * is added to the scene, and the renderer must decide how to render it.
   * It allows Passes to select geometries to handle the drawing of.
   * @param {TreeItem} treeItem - The treeItem value.
   * @param {object} rargs - Extra return values are passed back in this object.
   * The object contains a parameter 'continueInSubTree', which can be set to false,
   * so the subtree of this node will not be traversed after this node is handled.
   * @return {Boolean} - The return value.
   */
  itemAddedToScene(treeItem, rargs) {
    if (treeItem instanceof PointCloudAsset) {
      this.addPotreeasset(treeItem)
      return true
    }
    return false
  }

  /**
   * The itemRemovedFromScene method is called on each pass when aa item
   * is removed to the scene, and the pass must handle cleaning up any resources.
   * @param {TreeItem} treeItem - The treeItem value.
   * @param {object} rargs - Extra return values are passed back in this object.
   * @return {Boolean} - The return value.
   */
  itemRemovedFromScene(treeItem, rargs) {
    if (treeItem instanceof PointCloudAsset) {
      this.removePotreeasset(treeItem)
      return true
    }
    return false
  }

  /**
   * Adds a new point cloud asset the the list of assets rendered by this abstraction.
   *
   * @param {PointCloudAsset} pointcloudAsset - The pointcloudAsset value
   */
  addPotreeasset(pointcloudAsset) {
    const __bindAsset = (pointcloudAsset) => {
      const glpointcloudAsset = new GLPointCloudAsset(this.__gl, pointcloudAsset, this.glshader)
      glpointcloudAsset.on('updated', () => this.emit('updated'))
      pointcloudAsset.on('highlightChanged', () => {
        if (pointcloudAsset.isHighlighted()) {
          if (this.hilghlightedAssets.indexOf(glpointcloudAsset) == -1) this.hilghlightedAssets.push(glpointcloudAsset)
        } else {
          if (this.hilghlightedAssets.indexOf(glpointcloudAsset) != -1)
            this.hilghlightedAssets.splice(this.hilghlightedAssets.indexOf(glpointcloudAsset), 1)
        }
      })

      this.glpointcloudAssets.push(glpointcloudAsset)
    }
    if (pointcloudAsset.isLoaded()) __bindAsset(pointcloudAsset)
    else {
      pointcloudAsset.on('loaded', () => __bindAsset(pointcloudAsset))
    }
  }

  removePotreeasset(pointcloudAsset) {}

  // ///////////////////////////////////
  // Visibility
  /**
   * Sets viewport that is used by current abstraction.
   *
   * @param {GLViewport} viewport - The viewport value
   */
  setViewport(viewport) {
    this.viewport = viewport
    this.viewport.on('viewChanged', () => {
      this.visibleNodesNeedUpdating = true
    })
    this.viewport.on('resized', (event) => {
      const { width, height } = event

      this.viewportSize = [width, height]

      if (this.primaryRenderTarget) {
        this.primaryRenderTarget.resize(width, height)
      }
      if (this.smoothedRenderTarget) {
        this.smoothedRenderTarget.resize(width, height)
      }

      this.visibleNodesNeedUpdating = true
    })
    this.visibleNodesNeedUpdating = true
    this.viewportSize = [viewport.getWidth(), viewport.getHeight()]
  }

  /**
   * The updateVisibilityStructures method
   *
   * @param {array} priorityQueue - The priorityQueue value
   * @return {array} - The result
   */
  updateVisibilityStructures(priorityQueue) {
    const camera = this.viewport.getCamera()
    const view = camera.globalXfoParam.value.toMat4()
    const viewI = this.viewport.getViewMatrix()
    const proj = this.viewport.getProjectionMatrix()
    const viewProj = proj.multiply(viewI)
    const result = []
    this.glpointcloudAssets.forEach((glpointcloudAsset, index) => {
      const pointcloudAsset = glpointcloudAsset.getGeomItem()
      const model = pointcloudAsset.getGlobalMat4()
      const modelViewProj = viewProj.multiply(model)
      const frustum = new Frustum()
      frustum.setFromMatrix(modelViewProj)

      // camera  position in object space
      const modelInv = model.inverse()
      const camMatrixObject = modelInv.multiply(view)
      const camObjPos = camMatrixObject.translation

      if (pointcloudAsset.isVisible() && pointcloudAsset.pcoGeometry !== null) {
        priorityQueue.push({
          index,
          node: pointcloudAsset.pcoGeometry.root,
          weight: Number.MAX_VALUE,
        })
      }

      result.push({
        glpointcloudAsset,
        frustum,
        camObjPos,
      })
    })

    return result
  }

  /**
   * The updateVisibility method
   */
  updateVisibility() {
    const priorityQueue = new BinaryHeap(function (x) {
      return 1 / x.weight
    })
    const camera = this.viewport.getCamera()

    this.numVisiblePoints = 0
    let numVisiblePoints = 0
    const visibleNodesByAsset = []
    let visibleNodes = []
    const unloadedGeometry = []

    // calculate object space frustum and cam pos and setup priority queue
    const result = this.updateVisibilityStructures(priorityQueue)

    while (priorityQueue.size() > 0) {
      const element = priorityQueue.pop()
      const index = element.index
      const node = element.node

      if (numVisiblePoints + node.numPoints > this.pointBudget) {
        break
      }

      const frustum = result[index].frustum
      const insideFrustum = frustum.intersectsBox(node.boundingBox)
      if (!insideFrustum) {
        continue
      }
      numVisiblePoints += node.numPoints
      this.numVisiblePoints += node.numPoints

      const parent = element.parent
      if (!parent || parent.isLoaded()) {
        if (node.isLoaded()) {
          if (!visibleNodesByAsset[index]) visibleNodesByAsset[index] = []
          visibleNodesByAsset[index].push(node)

          visibleNodes.push(node)
        } else {
          unloadedGeometry.push(node)
        }
      }

      // add child nodes to priorityQueue
      const camObjPos = result[index].camObjPos
      const children = node.getChildren()
      const fov = camera.getFov()
      for (let i = 0; i < children.length; i++) {
        const child = children[i]

        let weight = 0
        if (true || camera.isPerspectiveCamera) {
          const sphere = child.getBoundingSphere()
          const distance = sphere.pos.distanceTo(camObjPos)
          const radius = sphere.radius
          if (distance - radius < 0) {
            weight = Number.MAX_VALUE
          } else {
            const slope = Math.tan(fov / 2)

            const projFactor = 0.5 / (slope * distance)
            const screenVRadius = radius * projFactor

            if (screenVRadius < this.minimumNodeVSize) {
              continue
            }
            weight = screenVRadius
          }
        } else {
          // TODO ortho visibility
          let bb = child.getBoundingBox()
          let distance = child.getBoundingSphere().pos.distanceTo(camObjPos)
          let diagonal = bb.max.clone().sub(bb.min).length()
          //weight = diagonal / distance;

          weight = diagonal
        }

        priorityQueue.push({
          index,
          node: child,
          parent: node,
          weight: weight,
        })
      }
    } // end priority queue loop

    const visibleNodeTextureOffsets = this.computeVisibilityTextureData(visibleNodes)

    visibleNodesByAsset.forEach((assetVisibleNodes, index) => {
      this.glpointcloudAssets[index].setVisibleNodes(assetVisibleNodes, this.lru, visibleNodeTextureOffsets)
    })

    if (unloadedGeometry.length > 0) {
      // Disabled temporarily
      // for (let i = 0; i < Math.min(Potree.maxNodesLoading, unloadedGeometry.length); i++) {
      const promises = []
      for (let i = 0; i < unloadedGeometry.length; i++) {
        // console.log("load:", unloadedGeometry[i].name);
        promises.push(unloadedGeometry[i].load())
      }
      if (promises.length > 0) {
        // After all the loads have finished.
        // update again so we can recompute and visiblity.
        Promise.all(promises).then(() => {
          // for (let i = 0; i < unloadedGeometry.length; i++) {
          //   console.log("loaded:", unloadedGeometry[i].name);
          // }
          this.visibleNodesNeedUpdating = true
          this.emit('updated')
        })
      }
    }

    // Causes unused nodes to be flushed.
    this.lru.freeMemory()

    // this.emit('updated');
  }

  /**
   * The computeVisibilityTextureData method
   *
   * @param {array} nodes - the nodes value
   * @return {Map} - The result
   */
  computeVisibilityTextureData(nodes) {
    const data = new Uint8Array(nodes.length * 4)
    const visibleNodeTextureOffsets = new Map()

    // copy array
    nodes = nodes.slice()

    // sort by level and index, e.g. r, r0, r3, r4, r01, r07, r30, ...
    const sort = function (a, b) {
      const na = a.name
      const nb = b.name
      if (na.length !== nb.length) return na.length - nb.length
      if (na < nb) return -1
      if (na > nb) return 1
      return 0
    }
    nodes.sort(sort)

    // const nodeMap = new Map();
    const offsetsToChild = new Array(nodes.length).fill(Infinity)

    for (let i = 0; i < nodes.length; i++) {
      const node = nodes[i]
      // nodeMap.set(node.name, node);
      visibleNodeTextureOffsets.set(node, i)

      if (i > 0) {
        const index = node.index //parseInt(node.name.slice(-1));
        // console.log(node.name, node.index, node.name.slice(-1))
        // const parentName = node.name.slice(0, -1);
        const parent = node.parent //nodeMap.get(parentName);
        // console.log(node.parent.name, parent.name, node.parent === parent)

        const parentIndex = visibleNodeTextureOffsets.get(parent)

        const parentOffsetToChild = i - parentIndex

        const childOffset = Math.min(offsetsToChild[parentIndex], parentOffsetToChild)

        // Add this bit to the parent's chils bit mask.
        data[parentIndex * 4 + 0] = data[parentIndex * 4 + 0] | (1 << index)
        data[parentIndex * 4 + 1] = childOffset >> 8
        data[parentIndex * 4 + 2] = childOffset % 256
        offsetsToChild[parentIndex] = childOffset
      }

      data[i * 4 + 3] = node.name.length - 1
    }

    this.visibleNodesTexture.populate(data, nodes.length, 1)
    return visibleNodeTextureOffsets
  }

  // ///////////////////////////////////
  // Rendering

  /**
   * The draw method.
   * @param {object} renderstate - The renderstate param.
   */
  draw(renderstate) {
    if (this.glpointcloudAssets.length == 0) return

    if (this.visibleNodesNeedUpdating) {
      this.updateVisibility()
      this.visibleNodesNeedUpdating = false
    }

    const gl = this.__gl

    gl.disable(gl.BLEND)
    gl.depthMask(true)
    gl.enable(gl.DEPTH_TEST)

    this.glshader.bind(renderstate)

    const { visibleNodes } = renderstate.unifs
    if (visibleNodes) this.visibleNodesTexture.bindToUniform(renderstate, visibleNodes)

    // gl.uniform1f(PointSize.location, this.pointSize);

    if (this.eyeDomeLighting) {
      if (!this.primaryRenderTarget) {
        this.initEyeDomeLighting()
      }

      this.primaryRenderTarget.bindForWriting(renderstate, true)

      // RENDER
      this.glpointcloudAssets.forEach((a) => a.draw(renderstate))
      this.primaryRenderTarget.unbindForWriting(renderstate)

      // SMOOTHING
      {
        this.smoothedRenderTarget.bindForWriting(renderstate, true)
        this.smoothingShader.bind(renderstate)

        const unifs = renderstate.unifs
        this.primaryRenderTarget.bindColorTexture(renderstate, unifs.colorTexture)

        if (unifs.viewportSize) gl.uniform2f(unifs.viewportSize.location, this.viewportSize[0], this.viewportSize[1])
        if (unifs.smoothRadius) gl.uniform1f(unifs.smoothRadius.location, 1.0)

        this.quad.bindAndDraw(renderstate)

        this.smoothedRenderTarget.unbindForWriting(renderstate)
      }

      // EYE DOME LIGHTING
      {
        this.eyeDomeLightingShader.bind(renderstate)

        const unifs = renderstate.unifs
        this.smoothedRenderTarget.bindColorTexture(renderstate, unifs.uEDLColor)

        gl.uniform2f(unifs.viewportSize.location, this.viewportSize[0], this.viewportSize[1])
        if (unifs.edlStrength) gl.uniform1f(unifs.edlStrength.location, 0.5)
        if (unifs.radius) gl.uniform1f(unifs.radius.location, 1.0)

        this.quad.bindAndDraw(renderstate)
      }
    } else {
      // RENDER
      this.glpointcloudAssets.forEach((a) => a.draw(renderstate))
    }
  }

  /**
   * The drawHighlightedGeoms method.
   * @param {object} renderstate - The renderstate param.
   */
  drawHighlightedGeoms(renderstate) {
    if (this.hilghlightedAssets.length == 0) return
    const gl = this.__gl

    gl.disable(gl.BLEND)
    gl.depthMask(true)
    gl.enable(gl.DEPTH_TEST)

    this.glhighlightShader.bind(renderstate)

    const { visibleNodes, PointSize } = renderstate.unifs
    if (visibleNodes) this.visibleNodesTexture.bindToUniform(renderstate, visibleNodes)

    gl.uniform1f(PointSize.location, this.pointSize)

    this.hilghlightedAssets.forEach((a) => a.drawHighlightedGeoms(renderstate))
  }

  /**
   * The drawGeomData method.
   * @param {object} renderstate - The renderstate param.
   */
  drawGeomData(renderstate) {
    if (this.glpointcloudAssets.length == 0) return
    const gl = this.__gl

    gl.disable(gl.BLEND)
    gl.depthMask(true)
    gl.enable(gl.DEPTH_TEST)

    this.glgeomdataShader.bind(renderstate)

    const { visibleNodes, PointSize } = renderstate.unifs
    if (visibleNodes) this.visibleNodesTexture.bindToUniform(renderstate, visibleNodes)

    gl.uniform1f(PointSize.location, this.pointSize)

    // RENDER
    this.glpointcloudAssets.forEach((a, index) => {
      const { passId, assetId } = renderstate.unifs
      if (passId) {
        gl.uniform1i(passId.location, this.passIndex)
      }
      if (assetId) {
        gl.uniform1i(assetId.location, index)
      }
      a.drawGeomData(renderstate)
    })
  }

  /**
   * The getGeomItemAndDist method.
   * @param {array} geomData - The geomData param.
   */
  getGeomItemAndDist(geomData) {
    const itemId = Math.round(geomData[1])
    const dist = geomData[3]
    const glpointcloudAsset = this.glpointcloudAssets[itemId]
    if (glpointcloudAsset) {
      return {
        geomItem: glpointcloudAsset.getGeomItem(),
        dist,
      }
    }
  }
}

export default GLPointCloudPass
export { GLPointCloudPass }
